### ISSUE TITLE
In the mac system, the case of the file under the storage is inconsistent in the Linux system.

### ISSUE THREAD

**Issue by jiandan217** [rails/rails#33864]
### Steps to reproduce
I use ActiveStorage to store images on disk.My rails project works fine under mac system, but the image can't be found when deployed to Linux environment.After debugging, the file name under storage/ is not consistent with the key case in the active_storage_blobs table. Although the storage name is inconsistent, you can find the image in the mac environment, not in linux.

For example

In database：
active_storage_blobs   key:2BANTEUdRXLbF1JWPKjNtD6i

In rails project:

Storage file structure

  storage
     ----**2b**
         ------AN
            --------2BANTEUdRXLbF1JWPKjNtD6i

The file name under the storage file is 2b.
Run the project in the Linux environment and get this error.
Disk Storage (0.5ms) Downloaded file from key: 2BANTEUdRXLbF1JWPKjNtD6i
Completed 500 Internal Server Error in 80ms (ActiveRecord: 0.0ms)
Errno::ENOENT (No such file or directory @ rb_sysopen - /home/project/bee/storage/**2B**/AN/2BANTEUdRXLbF1JWPKjNtD6i):

### System configuration
**Rails version**:
5.2.0
**Ruby version**:
2.5.1

**Comment by noahc** [comment]
Can you produce a small sample app that reproduces this problem?

**Comment by ur0** [comment]
@jiandan217 are you encountering this issue on a case-insensitive filesystem (e.g. case-insensitive HFS+)?

**Comment by jiandan217** [comment]
> @jiandan217 are you encountering this issue on a case-insensitive filesystem (e.g. case-insensitive HFS+)?

Thanks,
I found that the default system partition format for mac is journaled, and the file name is not case sensitive.

**Comment by julik** [comment]
Can this be reopened please? It actually is a problem in the following situation:

* An app is developed on macOS and hydrated with content
* The app is then released with the disk storage adapter. storage tree is synced to a linux server
* Some attached files then cannot be found due to different capitalisation of the partitioning subdirectory.

For example, on my application I get

```
{
  "errors": [
    "No such file or directory @ rb_sysopen - /homecoming-mk2/storage/mo/1m/mo1m9dFsf724WjzYYyQXNytM"
  ]
}
```

On the server:

```
root@<redacted>:~/<redacted>/Mo/1m# file mo1m9dFsf724WjzYYyQXNytM 
mo1m9dFsf724WjzYYyQXNytM: JPEG image data, JFIF standard 1.02, resolution (DPI), density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=7, orientation=upper-left, xresolution=98, yresolution=106, resolutionunit=2, software=Adobe Photoshop CS Macintosh, datetime=2005:04:06 16:03:36], progressive, precision 8, 110x110, frames 3
```

When I select the data for the blob in SQLite I get the row back:

```
sqlite> select * from active_storage_blobs where key = "mo1m9dFsf724WjzYYyQXNytM";
30|mo1m9dFsf724WjzYYyQXNytM|petrov.jpg|image/jpeg|{"identified":true,"analyzed":true}|25962|/A9UqZ0u8zgD7fpi3RzosA==|2018-12-25 23:24:40.569342
sqlite> \q
```

I need to look around to find out where this happens but at the very least the translation from the chunking subdirectory to the blob key must be deterministic, both ways. Otherwise it becomes impossible to transport the storage dir between filesystems, accessing the file becomes impossible as well. For me the error happens in code like this:

```ruby
blob = ActiveStorage::Blob.find(blob_id)
    blob.download do |bytes|
      tf.write(bytes)
    end
```

IMO just saying "But the Mac filesystem is broken" is not entirely right here. Afaik both HFS+ and AFS in the standard setup are case-insensitive and case-preserving.

**Comment by jiandan217** [comment]
> Can this be reopened please? It actually is a problem in the following situation:
> 
> * An app is developed on macOS and hydrated with content
> * The app is then released with the disk storage adapter. storage tree is synced to a linux server
> * Some attached files then cannot be found due to different capitalisation of the partitioning subdirectory.
> 
> For example, on my application I get
> 
> ```
> {
>   "errors": [
>     "No such file or directory @ rb_sysopen - /homecoming-mk2/storage/mo/1m/mo1m9dFsf724WjzYYyQXNytM"
>   ]
> }
> ```
> On the server:
> 
> ```
> root@<redacted>:~/<redacted>/Mo/1m# file mo1m9dFsf724WjzYYyQXNytM 
> mo1m9dFsf724WjzYYyQXNytM: JPEG image data, JFIF standard 1.02, resolution (DPI), density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=7, orientation=upper-left, xresolution=98, yresolution=106, resolutionunit=2, software=Adobe Photoshop CS Macintosh, datetime=2005:04:06 16:03:36], progressive, precision 8, 110x110, frames 3
> ```
> When I select the data for the blob in SQLite I get the row back:
> 
> ```
> sqlite> select * from active_storage_blobs where key = "mo1m9dFsf724WjzYYyQXNytM";
> 30|mo1m9dFsf724WjzYYyQXNytM|petrov.jpg|image/jpeg|{"identified":true,"analyzed":true}|25962|/A9UqZ0u8zgD7fpi3RzosA==|2018-12-25 23:24:40.569342
> sqlite> \q
> ```
> I need to look around to find out where this happens but at the very least the translation from the chunking subdirectory to the blob key must be deterministic, both ways. Otherwise it becomes impossible to transport the storage dir between filesystems, accessing the file becomes impossible as well. For me the error happens in code like this:
> 
> ```ruby
> blob = ActiveStorage::Blob.find(blob_id)
>     blob.download do |bytes|
>       tf.write(bytes)
>     end
> ```
> 
> IMO just saying "But the Mac filesystem is broken" is not entirely right here. Afaik both HFS+ and AFS in the standard setup are case-insensitive and case-preserving.
> 
> This is how the storage tree looks on macOS
> 
> ![capitalisation_activestorage](https://user-images.githubusercontent.com/16446/50459097-8094e080-0968-11e9-804e-855458e72071.png)

Sorry for the late reply,I have reopened this issues.Before this reply, I thought it was a system case setting problem. I see that you have created a new issues and have submitted the changes. Thank you for reminding me.

**Comment by georgeclaghorn** [comment]
This was fixed by #34818. :tada:
